File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Node {
    String w;
    int cnt;
    
    Node(String w, int cnt) {
        this.w = w;
        this.cnt = cnt;
    }
}

class Solution {
    String target;
    String[] words;
    boolean[] vis;
    
    boolean check(String a, String b) {
        int cnt = 0;
        // 다른게 오직 하나!
        for(int i=0; i<a.length(); i++) {
            char chA = a.charAt(i); char chB = b.charAt(i);
            if(chA != chB) cnt++;
        }
        
        return cnt == 1;
    }
    
    int bfs(String start) {
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(start, 0));
        
        while(!q.isEmpty()) {
            Node curr = q.poll();
            if(curr.w.equals(target)) {
                return curr.cnt;
            }
            
            for(int i=0; i<words.length; i++) {
                if(!vis[i] && check(curr.w, words[i])) {
                    vis[i] = true;
                    q.add(new Node(words[i], curr.cnt+1));
                }
            }
        }
        return 0;
    }
     
    public int solution(String begin, String target, String[] words) {
        this.target = target;
        this.words = words;
        
        vis = new boolean[words.length];
        return bfs(begin);
        
        // int answer = 0;
        // return answer;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;
int INF = 98765432;
int ans = INF;
string b, t;
vector<string> arr;
vector<bool> used;

bool check(string s1, string s2) {
    int cnt = 0;
    for(int i=0; i<s1.size(); i++) {
        if(s1[i] != s2[i]) cnt++;
    }
    return cnt==1;
}

void dfs(string curr, int cnt) {
    if(curr == t) {
        ans = min(ans, cnt);
        return;
    }
    if(ans < cnt) return;
    
    for(int i=0; i<arr.size(); i++) {
        if(!used[i] && check(curr, arr[i])) {
            used[i] = true;
            dfs(arr[i], cnt+1);
            used[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    b = begin; t = target; arr = words;
    used.resize(arr.size(), false);
    
    dfs(b, 0);
    if(ans == INF) return 0;
    else return ans;
}